### PR TITLE
[JENKINS-29601] Workflow compatibility

### PR DESCRIPTION
**Pull request by amuniz**
Downstream of https://github.com/jenkinsci/analysis-core-plugin/pull/31

https://issues.jenkins-ci.org/browse/JENKINS-29601

This changes allow to use ```FindbugsPublisher``` in a Workflow script, for example:

```
node {
  step([$class: 'FindBugsPublisher', pattern: '**/findbugsXml.xml', failedTotalAll: '0', usePreviousBuildAsReference: false])
}
```

### CODE REVIEW

**Comment by amuniz** [comment]
@reviewbybees

File: plugin/src/main/java/hudson/plugins/findbugs/FindBugsDescriptor.java [hunk 1]
```diff
@@ -1,6 +1,7 @@
 package hudson.plugins.findbugs;
 
 import hudson.Extension;
+
```
Comment (jtnord) [inline plugin/src/main/java/hudson/plugins/findbugs/FindBugsDescriptor.java]: nit - file has not been changed but has whitespace changes.

File: plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java [hunk 2]
```diff
@@ -154,6 +164,21 @@ public boolean isRankActivated() {
     }
 
     /**
+     * Added to properly uncoercing.
+     */
+    public boolean getIsRankActivated() {
+        return isRankActivated;
```
Comment (jtnord) [inline plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java]: this is the wrong name for this getter.
It should be `isIsRankActivated()` (sic).
I would recommend you rename this field so its called rankActivated - and then have a more sane name.

File: plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java [hunk 3]
```diff
@@ -154,6 +164,21 @@ public boolean isRankActivated() {
     }
 
     /**
+     * Added to properly uncoercing.
```
Comment (jtnord) [inline plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java]: useless javadoc.

Comment (amuniz) [inline plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java]: It was added because (I thought) it's not obvious that this method is only needed to have the workflow Snippetizer working fine (the method is not used anywhere in the project actually)

File: plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java [hunk 4]
```diff
@@ -180,19 +221,32 @@ public String getIncludePattern() {
...
     }
 
     @Override
-    public BuildResult perform(final AbstractBuild<?, ?> build, final PluginLogger logger) throws InterruptedException, IOException {
+    public BuildResult perform(final Run<?, ?> build, final FilePath workspace, final PluginLogger logger) throws InterruptedException, IOException {
         logger.log("Collecting findbugs analysis files...");
 
-        String defaultPattern = isMavenBuild(build) ? MAVEN_DEFAULT_PATTERN : ANT_DEFAULT_PATTERN;
+        String defaultPattern = ANT_DEFAULT_PATTERN;
+        boolean isMavenBuild = build instanceof AbstractBuild && isMavenBuild((AbstractBuild) build);
```
Comment (jtnord) [inline plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java]: just fix `isMavenBuild()` to take a `Run`.

File: plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        WorkflowJob job = j.jenkins.createProject(WorkflowJob.class, "wf");
+        File wfile = new File(j.jenkins.getRootDir() + "/workspace/wf/target");
+        wfile.mkdirs();
+        FilePath report = new FilePath(wfile).child("findbugs.xml");
+        report.copyFrom(WorkflowCompatibilityTest.class.getResourceAsStream("/hudson/plugins/findbugs/parser/findbugs-native.xml"));
+        job.setDefinition(new CpsFlowDefinition(
+        "node {" +
+        "  step([$class: 'FindBugsPublisher'])" +
+        "}"));
+        j.assertBuildStatusSuccess(job.scheduleBuild2(0));
```
Comment (jtnord) [inline plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java]: surely you want to report that the publisher picked up the file and reported the results?

File: plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+import hudson.model.Result;
+
+import hudson.plugins.findbugs.FindBugsPublisher;
+
+/**
+ * Test workflow compatibility.
+ */
+public class WorkflowCompatibilityTest {
+
+    @Rule
```
Comment (jtnord) [inline plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java]: ClassRule?

File: plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        wfile.mkdirs();
+        FilePath report = new FilePath(wfile).child("findbugsXml.xml");
+        report.copyFrom(WorkflowCompatibilityTest.class.getResourceAsStream("/hudson/plugins/findbugs/parser/findbugs-native.xml"));
+        job.setDefinition(new CpsFlowDefinition(
+        "node {" +
+        "  step([$class: 'FindBugsPublisher', pattern: '**/findbugsXml.xml', failedTotalAll: '0', usePreviousBuildAsReference: false])" +
+        "}"));
+        j.assertBuildStatus(Result.FAILURE, job.scheduleBuild2(0).get());
+    }
+}
```
Comment (jtnord) [inline plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java]: missing a test (& support?) for UNSTABLE builds?

File: plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    /**
+     * Run a workflow job using {@link FindBugsPublisher} with a failing threshols of 0, so the given example file
+     * "/hudson/plugins/findbugs/parser/findbugs-native.xml" will make the build to fail.
+     */
+    @Test
+    public void findbugsPublisherWorkflowStepSetLimits() throws Exception {
+        j.jenkins.getInjector().injectMembers(this);
+        WorkflowJob job = j.jenkins.createProject(WorkflowJob.class, "wf");
```
Comment (jtnord) [inline plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java]: use a classRule and a different job name?

File: plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java [hunk 9]
```diff
@@ -180,19 +221,32 @@ public String getIncludePattern() {
...
+    }
+
     @Override
     public Action getProjectAction(final AbstractProject<?, ?> project) {
         return new FindBugsProjectAction(project);
     }
 
     @Override
-    public BuildResult perform(final AbstractBuild<?, ?> build, final PluginLogger logger) throws InterruptedException, IOException {
+    public BuildResult perform(final Run<?, ?> build, final FilePath workspace, final PluginLogger logger) throws InterruptedException, IOException {
```
Comment (jtnord) [inline plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java]: I think this is wrong - you have changed the behavior of the class as it no longer takes into account the thresholds (due to the parent no longer calling this.
Or I'm missing something as the [parent](https://github.com/jenkinsci/analysis-core-plugin/blob/master/src/main/java/hudson/plugins/analysis/core/HealthAwarePublisher.java#L148-L176)  hasn't had any changes in master and there is no SNAPSHOT update in the POM.

Comment (jtnord) [inline plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java]: oh read the top level comment...

**Comment by jtnord** [comment]
a few too many issues for me so :bug:

File: plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java (hunk 2 above)
Comment (amuniz) [inline plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java]: I think the data migration is a risky process, we can live with this method name (unless you are absolutely against it).

File: plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java (hunk 9 above)
Comment (amuniz) [inline plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java]: Yep :-)

File: plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java (hunk 2 above)
Comment (jtnord) [inline plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java]: the method name is wrong.   the return is a boolean so it needs to start with `is` regardless of if you change the name of the bean property

File: plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java (hunk 4 above)
Comment (amuniz) [inline plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java]: It's like moving the ```instanceof``` inside ```isMavenBuild``` and as a reward we have another backward compatibility issue (since ```isMavenBuild``` is protected in the superclass so there could be more plugins out there overriding it).

File: plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java (hunk 6 above)
Comment (amuniz) [inline plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java]: What's the difference? I've never used ```ClassRule```.

File: plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java (hunk 5 above)
Comment (amuniz) [inline plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java]: I wanted to check that the build works fine, which is really good after all the changes!! :-)
Anyway, I'm going to improve this test to check what you propose.

File: plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java (hunk 8 above)
Comment (amuniz) [inline plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java]: Why a different job name? Each test re-build its own Jenkins instance from scratch (so the previous test should not be "visible" here).

File: plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java (hunk 7 above)
Comment (amuniz) [inline plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java]: Yeah, right. I'll add it.

File: plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java (hunk 4 above)
Comment (jtnord) [inline plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java]: you can (should be able to?) change the parameter type of `isMavenBuild` to a super type of the parameter without breaking backwards compatability.
Even if you can't adding a second method here would prevent this code being repeated in all the other static analysis plugins that also do a similar check.

File: plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java (hunk 6 above)
Comment (jtnord) [inline plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java]: a classRule will wrap around the class rather than test method - so the setup cost is only incurred once.

File: plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java (hunk 8 above)
Comment (jtnord) [inline plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java]: > "because each test gets its own jenkins instance"

That's why I'm saying use a `ClassRule` - you do not need this overhead per test.

which is then why you need a different job name :-)

**Comment by amuniz** [comment]
@jtnord your comments have been addressed

File: plugin/pom.xml [hunk 10]
```diff
@@ -68,6 +68,24 @@
       <artifactId>xercesImpl</artifactId>
       <version>2.11.0</version>
     </dependency>
+    <dependency>
+      <groupId>org.jenkins-ci.plugins.workflow</groupId>
+      <artifactId>workflow-job</artifactId>
+      <version>1.4</version>
```
Comment (jglick) [inline plugin/pom.xml]: Introduce a `workflow.version` property.

File: plugin/src/main/java/hudson/plugins/findbugs/FindBugsDescriptor.java (hunk 1 above)
Comment (jglick) [inline plugin/src/main/java/hudson/plugins/findbugs/FindBugsDescriptor.java]: Right, it is a good idea to look at your own diff to ensure that everything in it serves some purpose!

File: plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java (hunk 3 above)
Comment (jglick) [inline plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java]: Generally public getters are required in order for `f:checkbox` to properly round-trip configuration. It is possible this was actually broken before your patch: if you forget the public getter, the checkbox will just be quietly unchecked every time you open the form. `JenkinsRule.configRoundTrip` (or, for Workflow steps, `StepConfigTester`) is the way to verify that your forms are actually loading the same data they save.

File: plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+import static org.junit.Assert.*;
+
+/**
+ * Test workflow compatibility.
+ */
+public class WorkflowCompatibilityTest {
+
+    @ClassRule
+    public static JenkinsRule j = new JenkinsRule();
```
Comment (jglick) [inline plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java]: Except in very specialized cases, you want to use `@Rule` with `JenkinsRule`, not `@ClassRule`. (The exception would be if you are quite sure nothing in the test modifies any state in Jenkins, which is clearly not the case here.)

File: plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+    /**
+     * Run a workflow job using {@link FindBugsPublisher} and check for success.
+     */
+    @Test
+    public void findbugsPublisherWorkflowStep() throws Exception {
+        j.jenkins.getInjector().injectMembers(this);
```
Comment (jglick) [inline plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java]: What exactly do you think you are injecting here? Delete.

File: plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+        WorkflowJob job = j.jenkins.createProject(WorkflowJob.class, "wf");
+        File wfile = new File(j.jenkins.getRootDir() + "/workspace/wf/target");
+        wfile.mkdirs();
+        FilePath report = new FilePath(wfile).child("findbugs.xml");
```
Comment (jglick) [inline plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java]: Use `j.jenkins.getWorkspaceFor(job)`.

File: plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    /**
+     * Run a workflow job using {@link FindBugsPublisher} with a failing threshols of 0, so the given example file
+     * "/hudson/plugins/findbugs/parser/findbugs-native.xml" will make the build to fail.
+     */
+    @Test
+    public void findbugsPublisherWorkflowStepSetLimits() throws Exception {
+        j.jenkins.getInjector().injectMembers(this);
+        WorkflowJob job = j.jenkins.createProject(WorkflowJob.class, "wf2");
```
Comment (jglick) [inline plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java]: Note that when using `@Rule` you do not need to remember to pick distinct names.

File: plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+        File wfile = new File(j.jenkins.getRootDir() + "/workspace/wf2/target");
+        wfile.mkdirs();
+        FilePath report = new FilePath(wfile).child("findbugsXml.xml");
+        report.copyFrom(WorkflowCompatibilityTest.class.getResourceAsStream("/hudson/plugins/findbugs/parser/findbugs-native.xml"));
+        job.setDefinition(new CpsFlowDefinition(
+        "node {" +
+        "  step([$class: 'FindBugsPublisher', pattern: '**/findbugsXml.xml', failedTotalAll: '0', usePreviousBuildAsReference: false])" +
+        "}"));
+        j.assertBuildStatus(Result.FAILURE, job.scheduleBuild2(0).get());
```
Comment (jglick) [inline plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java]: This is a pretty weak test. It would pass if you just forgot the closing `}` in your script!

**Comment by jglick** [comment]
:bee: though there are some things that could be improved.

**Comment by svanoort** [comment]
:bee: pending backcompatibility fixes in analysis-core (from what I can see of the tests, it might be good to solidify them a bit though).

**Comment by amuniz** [comment]
@reviewbybees done

File: plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java (hunk 15 above)
Comment (amuniz) [inline plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java]: Fixed. Checking the number of findbugs annotations after the build.

Comment (jtnord) [inline plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java]: As per @jglick 
```
FindBugsResultAction result = job.getLastBuild().getAction(FindBugsResultAction.class);
assertTrue(result.getResult().getAnnotations().size() == 2);
```

**Comment by jtnord** [comment]
@amuniz this can not be @reviewbybees done as I had an existing bug open

**Comment by jtnord** [comment]
:bee:

**Comment by amuniz** [comment]
@jtnord oh, true. I didn't remember your bug. Thanks.

So re-calling it now: 

@reviewbybees done

**Comment by jglick** [comment]
:bee:

File: plugin/src/main/java/hudson/plugins/findbugs/FindBugsDescriptor.java [hunk 16]
```diff
@@ -11,6 +11,7 @@
  */
 @Extension(ordinal = 100)
 public final class FindBugsDescriptor extends PluginDescriptor {
+
```
Comment (uhafner) [inline plugin/src/main/java/hudson/plugins/findbugs/FindBugsDescriptor.java]: Remove.

File: plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java [hunk 17]
```diff
@@ -142,6 +147,11 @@ public FindBugsPublisher(final String healthy, final String unHealthy, final Str
     }
     // CHECKSTYLE:ON
 
+    @DataBoundConstructor
```
Comment (uhafner) [inline plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java]: JavaDoc.

File: plugin/pom.xml [hunk 18]
```diff
@@ -45,7 +49,7 @@
     <dependency>
       <groupId>org.jvnet.hudson.plugins</groupId>
       <artifactId>analysis-core</artifactId>
-      <version>1.73-SNAPSHOT</version>
+      <version>1.73-beta-SNAPSHOT</version><!-- TODO: change to 1.73 before release -->
```
Comment (uhafner) [inline plugin/pom.xml]: why is this required?

File: plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java [hunk 19]
```diff
@@ -117,10 +120,12 @@
      *            RegEx patterns of files to exclude from the report
      * @param includePattern
      *            RegEx patterns of files to include in the report
+     * @deprecated This constructor is called internally only, but if you need to use it (for some strange reason), call
+     *            {@link #FindBugsPublisher()} and available setters
      */
     // CHECKSTYLE:OFF
     @SuppressWarnings("PMD.ExcessiveParameterList")
-    @DataBoundConstructor
+    @Deprecated
```
Comment (uhafner) [inline plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java]: You can delete  @Deprecated methods, this plug-in is not designed and used for extension.

File: plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java [hunk 20]
```diff
@@ -180,19 +221,32 @@ public String getIncludePattern() {
...
-    public BuildResult perform(final AbstractBuild<?, ?> build, final PluginLogger logger) throws InterruptedException, IOException {
+    public BuildResult perform(final Run<?, ?> build, final FilePath workspace, final PluginLogger logger) throws InterruptedException, IOException {
         logger.log("Collecting findbugs analysis files...");
 
-        String defaultPattern = isMavenBuild(build) ? MAVEN_DEFAULT_PATTERN : ANT_DEFAULT_PATTERN;
+        String defaultPattern = ANT_DEFAULT_PATTERN;
+        boolean isMavenBuild = isMavenBuild(build);
+        if (isMavenBuild) {
+            defaultPattern = MAVEN_DEFAULT_PATTERN;
+        }
```
Comment (uhafner) [inline plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java]: Why did you change the one-liner?

File: plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java [hunk 21]
```diff
@@ -202,6 +256,8 @@ public BuildResult perform(final AbstractBuild<?, ?> build, final PluginLogger l
         return result;
     }
 
+
+
```
Comment (uhafner) [inline plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java]: remove

File: plugin/src/main/java/hudson/plugins/findbugs/FindBugsResult.java [hunk 22]
```diff
@@ -155,4 +157,23 @@ public String getDisplayName() {
     protected Class<? extends ResultAction<? extends BuildResult>> getResultActionType() {
         return FindBugsResultAction.class;
     }
+
+    /**
+     * @deprecated use {@link #FindBugsResult(Run, String, ParserResult, boolean, boolean)} instead
+     */
+    @Deprecated
```
Comment (uhafner) [inline plugin/src/main/java/hudson/plugins/findbugs/FindBugsResult.java]: No deprecated code in this plugin.

File: plugin/src/main/java/hudson/plugins/findbugs/FindBugsResultAction.java [hunk 23]
```diff
@@ -40,4 +42,12 @@ public String getDisplayName() {
     protected PluginDescriptor getDescriptor() {
         return new FindBugsDescriptor();
     }
+
+    /**
+     * @deprecated use {@link #FindBugsResultAction(Run, HealthDescriptor, FindBugsResult)}
+     */
+    @Deprecated
```
Comment (uhafner) [inline plugin/src/main/java/hudson/plugins/findbugs/FindBugsResultAction.java]: delete

File: plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java [hunk 24]
```diff
@@ -0,0 +1,81 @@
...
+        FilePath workspace = j.jenkins.getWorkspaceFor(job);
+        FilePath report = workspace.child("target").child("findbugsXml.xml");
+        report.copyFrom(WorkflowCompatibilityTest.class.getResourceAsStream("/hudson/plugins/findbugs/parser/findbugs-native.xml"));
+        job.setDefinition(new CpsFlowDefinition(
+        "node {" +
+        "  step([$class: 'FindBugsPublisher', pattern: '**/findbugsXml.xml', failedTotalAll: '0', usePreviousBuildAsReference: false])" +
+        "}"));
+        j.assertBuildStatus(Result.FAILURE, job.scheduleBuild2(0).get());
+        FindBugsResultAction result = job.getLastBuild().getAction(FindBugsResultAction.class);
+        assertTrue(result.getResult().getAnnotations().size() == 2);
```
Comment (uhafner) [inline plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java]: Don't use assertTrue for integers: use assertEquals.

File: plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java [hunk 25]
```diff
@@ -0,0 +1,81 @@
...
+        FilePath workspace = j.jenkins.getWorkspaceFor(job);
+        FilePath report = workspace.child("target").child("findbugsXml.xml");
+        report.copyFrom(WorkflowCompatibilityTest.class.getResourceAsStream("/hudson/plugins/findbugs/parser/findbugs-native.xml"));
+        job.setDefinition(new CpsFlowDefinition(
+        "node {" +
+        "  step([$class: 'FindBugsPublisher', pattern: '**/findbugsXml.xml', unstableTotalAll: '0', usePreviousBuildAsReference: false])" +
+        "}"));
+        j.assertBuildStatus(Result.UNSTABLE, job.scheduleBuild2(0).get());
+        FindBugsResultAction result = job.getLastBuild().getAction(FindBugsResultAction.class);
+        assertTrue(result.getResult().getAnnotations().size() == 2);
```
Comment (uhafner) [inline plugin/src/test/java/hudson/plugins/findbugs/workflow/WorkflowCompatibilityTest.java]: assertEquals

**Comment by uhafner** [comment]
It would be nice if we could have an acceptance test with workflow job after this PR has been integerated...

File: plugin/pom.xml (hunk 18 above)
Comment (amuniz) [inline plugin/pom.xml]: All changes made on [analysis core PR](https://github.com/jenkinsci/analysis-core-plugin/pull/31) (upstream of this one) have been deployed as `1.73-beta-SNAPSHOT`, so it will not conflict with the real 1.73-SNAPSHOT (which should be linked to master branch, not from a PR).

In this way all downstream PRs compile successfully (in local environments and CI service) without requiring to checkout and compile locally the upstream PR in analysis-core. Once the upstream PR is merged in master, we can deploy the real 1.73-SNAPSHOT, then update all downstream PRs and merge then securely.

File: plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java (hunk 19 above)
Comment (amuniz) [inline plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java]: I was doing it as a rule, as we don't have real information about non-public plugins extending this one.
I would keep it, but you are the maintainer, so if it's fine for you I can delete it. Please, confirm.

File: plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java (hunk 20 above)
Comment (amuniz) [inline plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java]: To avoid two consecutive calls to `isMavenBuild` (its result value is used below again, line 247).

**Comment by amuniz** [comment]
@uhafner all your comments have been addressed. Thanks!

I wrote a file inline message about deprecated methods. What do you think?

**Comment by amuniz** [comment]
I can add the acceptance test later, of course. Added to my TODO list.

File: plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java (hunk 19 above)
Comment (uhafner) [inline plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java]: Yes, please delete deprecated methods. This makes the plug-in code too polluted...

Comment (amuniz) [inline plugin/src/main/java/hudson/plugins/findbugs/FindBugsPublisher.java]: Ok. Going for it.

**Comment by amuniz** [comment]
@uhafner Deprecated signatures removed

**Comment by uhafner** [comment]
While browsing the code in IntelliJ now I wondered why FindBugsMavenResultAction has not been adapted? I think the new code should be used there, too (AbstractBuild should be cast to Run).

**Comment by amuniz** [comment]
AFAICT this class is only used for Maven Job type, so the build object is granted to be an instance of `AbstractBuild`, that's why it was not adapted. It's not affecting in any way to the behavior of neither Workflow jobs or Maven jobs.

Now, I think it would be good to stop using the deprecated constructors of `FindbugsResult`and `FindbugsResultAction` even not impacting in any behavior. I'll do the changes.

**Comment by amuniz** [comment]
Indeed, after looking on it, it was not using deprecated constructors since they were removed. I only changed `FindBugsMavenResultAction` to use `Run` instead of `AbstractBuild`. As said before, an innocuous change.
